### ISSUE TITLE
Distributed tensorflow hang when turning from Async to Sync

### ISSUE THREAD

**Issue by smartcat2010** [tensorflow/tensorflow#2386]
I follow [Distributed tensorflow](https://www.tensorflow.org/versions/r0.8/how_tos/distributed/index.html) to port our RNN to multi-GPU cards. I start 4 processes and each process run on a different GPU card.
After solving the [issue](https://github.com/tensorflow/tensorflow/issues/2292), I could run it on my 4 GPU cards server in Asynchronized mode. 

Then I change my code to Synchronized mode by adding only the following code:
    optimizer = tf.train.SyncReplicasOptimizer(
        optimizer,
        replicas_to_aggregate = 4, 
        replica_id = FLAGS.task_index,
        total_num_replicas = 4
        #variable_averages = exp_moving_averager,    # Remove this for simple.
        #variables_to_average = variables_to_average # Remove this for simple.
        )

Then the job will hang in session.run(...)
Is there any other thing needed to change when switch from Async mode to Sync mode?

Thanks a lot in advance!

**Comment by jmchen-g** [comment]
You also need to start the chief queue runner which will start the synchronization thread. Please follow the comments and example in the sync_replicas_optimizer.py. Another example is the inception_distributed_train.py in tensorflow/models/inception

**Comment by smartcat2010** [comment]
@jmchen-g 
Thanks a lot for your help! I turn to [sync_replicas_optimizer.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/sync_replicas_optimizer.py) and follow the "Usage". In order to simplify the code, I just change [mnist_softmax.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/tutorials/mnist/mnist_softmax.py) to changing input data to random data. 
I start 1 ps and 2 workers in the same machine(it has 4 GPU cards). The worker 0 runs smoothly and print out many lines of "accuracy=XXX" as expected. But the worker 1 is crashed with the following error:

`I tensorflow/core/common_runtime/gpu/gpu_device.cc:755] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K40c, pci bus id: 0000:02:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:755] Creating TensorFlow device (/gpu:1) -> (device: 1, name: Tesla K40c, pci bus id: 0000:04:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:755] Creating TensorFlow device (/gpu:2) -> (device: 2, name: Tesla K40c, pci bus id: 0000:84:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:755] Creating TensorFlow device (/gpu:3) -> (device: 3, name: Tesla K40c, pci bus id: 0000:85:00.0)
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:206] Initialize HostPortsGrpcChannelCache for job ps -> {10.141.33.61:2222}
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:206] Initialize HostPortsGrpcChannelCache for job worker -> {10.141.33.61:2223, localhost:2224}
I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:202] Started server with target: grpc://localhost:2224
Traceback (most recent call last):
  File "./mnist_softmax.py", line 87, in <module>
    tf.app.run()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "./mnist_softmax.py", line 84, in main
    run_training(server, cluster)
  File "./mnist_softmax.py", line 72, in run_training
    _, cost, acc, step = sess.run([train_step, cross_entropy, accuracy, global_step], feed_dict = { x: source_data, y_ : labels_one_hot })
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 340, in run
    run_metadata_ptr)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 564, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 637, in _do_run
    target_list, options, run_metadata)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 659, in _do_call
    e.code)
tensorflow.python.framework.errors.InternalError: No worker known as /job:worker/replica:0/task:1
	 [[Node: sync_replicas/cond/switch_f_S24 = _Recv[client_terminated=false, recv_device="/job:ps/replica:0/task:0/cpu:0", send_device="/job:worker/replica:0/task:1/gpu:1", send_device_incarnation=-240247217847712065, tensor_name="edge_42_sync_replicas/cond/switch_f", tensor_type=DT_FLOAT, _device="/job:ps/replica:0/task:0/cpu:0"]()]]
`

The command line of worker1:
`python ./mnist_softmax.py --ps_hosts=10.141.33.61:2222 --worker_hosts=10.141.33.61:2223,10.141.33.61:2224 --job_name=worker --task_index=1 --gpu_cards=2`

All of my code:
`from __future__ import absolute_import
from __future__ import division
from __future__ import print_function


import numpy
import tensorflow as tf

tf.app.flags.DEFINE_string("ps_hosts", "localhost:2222", "...")
tf.app.flags.DEFINE_string("worker_hosts", "localhost:2223", "...")
tf.app.flags.DEFINE_string("job_name", "", "...")
tf.app.flags.DEFINE_integer("task_index", 0, "...")
tf.app.flags.DEFINE_integer('gpu_cards', 1, 'Number of GPU cards in a machine to use.')
FLAGS = tf.app.flags.FLAGS

def dense_to_one_hot(labels_dense, num_classes = 10) :
    """Convert class labels from scalars to one-hot vectors."""
    num_labels = labels_dense.shape[0]
    index_offset = numpy.arange(num_labels) * num_classes
    labels_one_hot = numpy.zeros((num_labels, num_classes))
    labels_one_hot.flat[index_offset + labels_dense.ravel()] = 1
    return labels_one_hot

def run_training(server, cluster):
    is_chief = (FLAGS.task_index == 0)
    with tf.Graph().as_default():        
        with tf.device(tf.train.replica_device_setter(
                worker_device = "/job:worker/task:%d" % FLAGS.task_index,
                cluster = cluster)) :
            with tf.device('/cpu:0') :
                global_step = tf.get_variable('global_step', [],
                    initializer = tf.constant_initializer(0), trainable = False)
            with tf.device('/gpu:%d' % (FLAGS.task_index % FLAGS.gpu_cards)) :                            
                # Create the model
                x = tf.placeholder("float", [None, 784])
                W = tf.Variable(tf.zeros([784, 10]))
                b = tf.Variable(tf.zeros([10]))
                y = tf.nn.softmax(tf.matmul(x, W) + b)

                # Define loss and optimizer
                y_ = tf.placeholder("float", [None, 10])
                cross_entropy = -tf.reduce_sum(y_ * tf.log(y))
                opt = tf.train.GradientDescentOptimizer(0.01)
                num_workers = FLAGS.gpu_cards # Use only 1 worker for test.
                opt = tf.train.SyncReplicasOptimizer(opt, replicas_to_aggregate = num_workers,
                    replica_id = FLAGS.task_index, total_num_replicas = num_workers)
                train_step = opt.minimize(cross_entropy, global_step = global_step)

                # Test trained model
                correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
                accuracy = tf.reduce_mean(tf.cast(correct_prediction, "float"))

            init_token_op = opt.get_init_tokens_op()
            chief_queue_runner = opt.get_chief_queue_runner()

            init = tf.initialize_all_variables()
            sv = tf.train.Supervisor(is_chief = is_chief,
                init_op = init,
                global_step = global_step)
            # Create a session for running Ops on the Graph.
            config = tf.ConfigProto(allow_soft_placement = True)
            sess = sv.prepare_or_wait_for_session(server.target, config = config)

            if is_chief:
                sv.start_queue_runners(sess, [chief_queue_runner])                
                sess.run(init_token_op)

            for i in range(100000):
                source_data = numpy.random.normal(loc = 0.0, scale = 1.0, size = (100, 784))
                labels_dense = numpy.clip(numpy.sum(source_data, axis = 1) / 5 + 5, 0, 9).astype(int)
                labels_one_hot = dense_to_one_hot(labels_dense)
                _, cost, acc, step = sess.run([train_step, cross_entropy, accuracy, global_step], feed_dict = { x: source_data, y_ : labels_one_hot })
                print("[%d]: cost=%.2f, accuracy=%.2f" % (step, cost, acc))


def main(_) :
    ps_hosts = FLAGS.ps_hosts.split(",")
    worker_hosts = FLAGS.worker_hosts.split(",")
    cluster = tf.train.ClusterSpec({ "ps": ps_hosts, "worker" : worker_hosts })
    server = tf.train.Server(cluster, job_name = FLAGS.job_name, task_index = FLAGS.task_index)
    if FLAGS.job_name == "ps":
      server.join()
    elif FLAGS.job_name == "worker" :
      run_training(server, cluster)

if __name__ == '__main__' :
  tf.app.run()
`

**Comment by smartcat2010** [comment]
@jmchen-g 
I've also follow the style of [inception_distributed_train.py](https://github.com/tensorflow/models/blob/master/inception/inception/inception_distributed_train.py) , and changed my above code to:
with tf.device('/job:worker/task:%d' % FLAGS.task_index) :

Then restarting the job, still 1 ps and 2 workers. Then worker 0 still runs smoothly and print out results. Worker 1 hangs inner "sv.prepare_or_wait_for_session"

This is my code: (Adding some "print" to locate where to hang)
[mnist_softmax.py.txt](https://github.com/tensorflow/tensorflow/files/269636/mnist_softmax.py.txt)

**Comment by jmchen-g** [comment]
Actually the inception distributed train is using slim which has its own trick related to device setter. For you case, can you try follow the instructions for the replica_device_setter:
  # To build a cluster with two ps jobs on hosts ps0 and ps1, and 3 worker
  # jobs on hosts worker0, worker1 and worker2.
  cluster_spec = {
      "ps": ["ps0:2222", "ps1:2222"],
      "worker": ["worker0:2222", "worker1:2222", "worker2:2222"]}
  with tf.device(tf.replica_device_setter(cluster=cluster_spec)):
    # Build your graph
    v1 = tf.Variable(...)  # assigned to /job:ps/task:0
    v2 = tf.Variable(...)  # assigned to /job:ps/task:1
    v3 = tf.Variable(...)  # assigned to /job:ps/task:0

**Comment by smartcat2010** [comment]
@jmchen-g 
Thank you very much for your quick help!
I updated my code following your guide. I run 1 ps and 2 workers like this:
```
cluster_spec = tf.train.ClusterSpec({ "ps":["10.141.33.61:2222"], "worker" : ["10.141.33.61:2223", "10.141.33.65:2223"] })
with tf.device(tf.train.replica_device_setter(cluster = cluster_spec)) :
```
My code file:
[mnist_softmax.py.txt](https://github.com/tensorflow/tensorflow/files/269861/mnist_softmax.py.txt)

Then worker0 runs smoothly, worker1 shows the following error:

```
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:206] Initialize HostPortsGrpcChannelCache for job ps -> {10.141.33.61:2222}
I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:206] Initialize HostPortsGrpcChannelCache for job worker -> {10.141.33.61:2223, localhost:2223}
I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:202] Started server with target: grpc://localhost:2223
Traceback (most recent call last):
  File "./mnist_softmax.py", line 83, in <module>
    tf.app.run()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "./mnist_softmax.py", line 80, in main
    run_training(server, cluster_spec)
  File "./mnist_softmax.py", line 70, in run_training
    _, cost, acc, step = sess.run([train_step, cross_entropy, accuracy, global_step], feed_dict = { x: source_data, y_ : labels_one_hot })
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 340, in run
    run_metadata_ptr)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 564, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 637, in _do_run
    target_list, options, run_metadata)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 659, in _do_call
    e.code)
tensorflow.python.framework.errors.InvalidArgumentError: Expected size[0] in [0, 0], but got 1
	 [[Node: get_local_step = Slice[Index=DT_INT32, T=DT_FLOAT, _device="/job:worker/replica:0/task:0/gpu:0"](local_steps_S11, Reshape, get_local_step/size)]]
Caused by op u'get_local_step', defined at:
  File "./mnist_softmax.py", line 83, in <module>
    tf.app.run()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "./mnist_softmax.py", line 80, in main
    run_training(server, cluster_spec)
  File "./mnist_softmax.py", line 45, in run_training
    train_step = opt.minimize(cross_entropy, global_step = global_step)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/training/optimizer.py", line 192, in minimize
    name=name)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/training/sync_replicas_optimizer.py", line 334, in apply_gradients
    name="get_local_step")
  File "/usr/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 217, in slice
    return gen_array_ops._slice(input_, begin, size, name=name)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1318, in _slice
    name=name)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/ops/op_def_library.py", line 655, in apply_op
    op_def=op_def)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2154, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1154, in __init__
    self._traceback = _extract_stack()
```

When I change it to run the 2 workers in the same machine as follows, then both of the workers runs smoothly.
`cluster_spec = tf.train.ClusterSpec({ "ps":["10.141.33.61:2222"], "worker" : ["10.141.33.61:2223", "10.141.33.61:2224"] })`

So is there something wrong in my multi-machine code?

**Comment by jmchen-g** [comment]
Can you remove all the checkpoint before starting the run?

If you want to have checkpoint, you need to make sure the path is
accessible from all PS and chief worker.

**Comment by smartcat2010** [comment]
@jmchen-g 
Do you mean the file with name "checkpoint"? I've just deleted all files including that under the code's folder of both machines. But the same error still shows.

**Comment by jmchen-g** [comment]
Looks like you only specified 1 worker? In your case, they should be 2 instead.

                **num_workers** = FLAGS.gpu_cards # Use only 1 worker for test.
                opt = tf.train.SyncReplicasOptimizer(opt, replicas_to_aggregate = num_workers,
                    replica_id = FLAGS.task_index, total_num_replicas = num_workers)

**Comment by smartcat2010** [comment]
@jmchen-g 
Thank you very much! It's my stupid mistake of specifying 1 worker in previous code. Now it works on multi-nodes/multi-cards after fixing my bug. Really sorry for bring so much trouble to you.
My fixed code which works:
[mnist_softmax.py.txt](https://github.com/tensorflow/tensorflow/files/271770/mnist_softmax.py.txt)

Maybe it's a good idea to put the usage of Sync-mode into [tutorial](https://www.tensorflow.org/versions/r0.8/how_tos/distributed/index.html). The current tutorial only has a Async-mode code example. Although Inception example code uses Sync-mode, it also uses more complex things such as slim, queues, so it's not easy for junior user to start.

Thanks a lot again for your kindly help!

**Comment by jmchen-g** [comment]
No problem. Glad it worked~:)

Ya, we might add an example at some point. Thanks for the suggestion.

**Comment by s0okiym** [comment]
I tried the newest mnist_softmax.py.txt provided above.
It seems hang at sv.prepare_or_wait_for_session in worker 1.

**Comment by s0okiym** [comment]
I tried both [Distributed tensorflow](https://www.tensorflow.org/versions/r0.8/how_tos/distributed/index.html)   and [minist_softmax.py.txt](https://github.com/tensorflow/tensorflow/files/271770/mnist_softmax.py.txt).  They both hang at the prepare_or_wait_for_session() function with non-chief worker tasks. 

task_index == 0(chief worker ) can run as expected.
I can not find out the reasion why it behaves like this. 
Can anyone please give some advice?

**Comment by mrry** [comment]
@s0okiym: It sounds like you might be having a different problem. Please open a new issue and describe exactly how you are invoking the various binaries.

**Comment by DjangoPeng** [comment]
@s0okiym 
I have the same problem!
only the chief task works, and the ps prints a list of errors 
"E tensorflow/stream_executor/cuda/cuda_driver.cc:965] failed to allocate 564.97M (592413952 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY"
Have you fixed your problem?

**Comment by icklerly** [comment]
@smartcat2010 
Hi,
I tried your mnist_softmax.py.txt example and did run it on 1 parameter server and 2 workers.
Unfortunately, I get a two times longer runtime with using 2 workers compared to using only 1 worker.
Can you confirm this behavior?
A feedback would be great!

**Comment by hbhuang** [comment]
@DjangoPeng @icklerly  
Hi all. I also meet the problem when i run mnist_softmax.py.txt on  1 parameter server and 2 workers. only the chief task works, and the ps prints a list of errors . 
can you help me?

**Comment by renato2099** [comment]
Did anybody got any answers or hints on this one?

**Comment by jiayiliu** [comment]
@jmchen-g Is there any changes with v1.3?
I tried to run with the MonitoredTrainingSession(), but the process seemed to hang in the middle.
Basically I followed the asynchronized tutorial (works fine) and add the SyncReplicasOptimizer.
I tried to play with the `get_init_tokens_op()` but no luck..
Any hints are welcome.  Thanks!

Here is my sample code and outputs:

`import tensorflow as tf
import sys

cluster_spec = {"ps":["localhost:8864"],
                 "worker":["localhost:8865","localhost:8866"]}


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("test.py ps|worker 0|1")
        exit(1)

    role = sys.argv[1]
    task_id = int(sys.argv[2])

    cluster = tf.train.ClusterSpec(cluster_spec)
    server = tf.train.Server(cluster, job_name=role, task_index=task_id)

    if role == "ps":
        server.join()
    else:
        with tf.device(
                tf.train.replica_device_setter(cluster=cluster_spec,
                    worker_device="/job:worker/task:%d"%task_id)):
            global_steps = tf.contrib.framework.get_or_create_global_step()
            # a very simple model with pseudo data
            inputs = tf.placeholder(tf.float32, shape=[None, 3])
            labels = tf.placeholder(tf.float32, shape=[None])
            w = tf.Variable([0.11,10,0.3])
            y = tf.reduce_sum(tf.multiply(inputs, w), axis=1)
            loss = tf.losses.mean_squared_error(labels, y)
            opt = tf.train.GradientDescentOptimizer(0.00001)
            opt_sync = tf.train.SyncReplicasOptimizer(opt,
                                                      replicas_to_aggregate=2,
                                                      # total_num_replicas=2,
                                                      use_locking=True)
            hooks = [opt_sync.make_session_run_hook((task_id==0))]
            op = opt_sync.minimize(loss, global_step=global_steps)

        with tf.train.MonitoredTrainingSession(master=server.target,
                                               is_chief=(task_id==0),
                                               hooks=hooks) as sess:
            for i in range(10):
                j = sess.run(global_steps)
                t = sess.run(loss, feed_dict={inputs:[[1,2,3]],labels:[1]})
                sess.run(op, feed_dict={inputs:[[1,2,3],[2,3,1]],labels:[2,1]})
                print("%d %d %f"%(i,j,t))

        print("finished")`

from worker 1:
`2017-10-03 15:48:37.079376: I tensorflow/core/distributed_runtime/master_session.cc:998] Start master session eb05b8a25a96918e with config: 
0 0 400.400116
1 1 400.16366`

from worker 2:
`2017-10-03 15:48:3 6.882892: I tensorflow/core/distributed_runtime/master_session.cc:998] Start master session e54098121f5f8494 with config: 
0 0 400.400116
1 1 400.400116`

**Comment by dingevin** [comment]
@jiayiliu Have you solve this problem?  the chief worker is hang when i use tf.train.SyncReplicasOptimizer and tf.train.MonitoredTrainingSession, but the other worker was finished and exit successfully. then the chief worker show as follow:
Exception in thread QueueRunnerThread-dummy_queue-sync_token_q_EnqueueMany:
      CancelledError: Step was cancelled by an explicit call to `Session::Close()`.

**Comment by YesterdayxD** [comment]
@dingevin  I have the same problem that you describe.Have you solve this problem?
